File: content/LLD/Reference-Code/12. Factory Pattern/good_example.py
from abc import ABC, abstractmethod


class Food(ABC):
    @abstractmethod
    def prepare(self):
        pass


class Pizza:
    def prepare(self):
        print("Preparing pizza")


class Burger:
    def prepare(self):
        print("Preparing burger")


class Pasta:
    def prepare(self):
        print("Preparing burger")


class FoodFactory:
    @staticmethod
    def create_food(food_type: str) -> Food:
        if food_type == "pizza":
            return Pizza()
        elif food_type == "burger":
            return Burger()
        elif food_type == "pasta":
            return Pasta()
        else:
            return None


class RestrauntService:
    def create_order(self, food_type: str):
        f = FoodFactory.create_food(food_type)
        if f is None:
            print("Cannot prepare food")
            return None
        f.prepare()
        return f


restraunt_service = RestrauntService()
restraunt_service.create_order("pizza")
restraunt_service.create_order("burger")

File: content/LLD/Reference-Code/12. Factory Pattern/GoodExample.java
public class GoodExample {
    interface Food {
        void prepare();
    }

    static class Pizza implements Food {
        @Override
        public void prepare() {
            System.out.println("Preparing pizza");
        }
    }

    static class Burger implements Food {
        @Override
        public void prepare() {
            System.out.println("Preparing burger");
        }
    }

    static class Pasta implements Food {
        @Override
        public void prepare() {
            System.out.println("Preparing pasta");
        }
    }

    static class FoodFactory {
        public static Food createFood(String foodType) {
            if ("pizza".equalsIgnoreCase(foodType)) {
                return new Pizza();
            } else if ("burger".equalsIgnoreCase(foodType)) {
                return new Burger();
            } else if ("pasta".equalsIgnoreCase(foodType)) {
                return new Pasta();
            }
            return null;
        }
    }

    static class RestaurantService {
        public Food createOrder(String foodType) {
            Food f = FoodFactory.createFood(foodType);
            if (f == null) {
                System.out.println("Cannot prepare food");
                return null;
            }
            f.prepare();
            return f;
        }
    }

    public static void main(String[] args) {
        RestaurantService restaurantService = new RestaurantService();
        restaurantService.createOrder("pizza");
        restaurantService.createOrder("burger");
    }
}
